docs: update README to describe Capacity Planner and GPU Recommender (#182)
redhat-et/compass: README.md

[![Python 3.11+](https://img.shields.io/badge/Python-3.11+-blue.svg)](https://www.python.org)
[![License: Apache 2.0](https://img.shields.io/badge/License-Apache_2.0-blue.svg)](LICENSE)

**From business requirements to running llm-d deployments — sizing, estimation, and deployment in one workflow.**

---

## Overview

Deploying LLMs in production means navigating a fragmented landscape of model selection, GPU sizing, performance estimation, and Kubernetes configuration. Platform teams typically piece together separate tools and manual calculations, leading to expensive trial-and-error.

llm-d Planner unifies these concerns into a single platform with three capabilities:

- **Planner** — The main workflow. Guides users from business requirements to a running LLM deployment through conversational requirements gathering, SLO-driven recommendations, and one-click Kubernetes deployment.

> - **Capacity Analyzer** — Estimate GPU memory requirements for any HuggingFace model: model weights, KV cache, activation memory, and CUDA overhead. Determine minimum GPU count, maximum context length, and concurrent request capacity for a given hardware configuration.

Reviewer: ```suggestion
- **Capacity Analyzer** — Estimate GPU memory requirements for any HuggingFace model: model weights, KV cache, activation memory, and system overhead. Determine minimum GPU count, maximum context length, and concurrent request capacity for a given hardware configuration.
```
Keep phrasing non-vendor specific.

Reviewer 2: Done.

- **Performance Analyzer** — Estimate inference performance (TTFT, ITL, throughput) across GPU types without running actual benchmarks. Compare latency and performance tradeoffs to find the optimal GPU for your workload.

Planner is the main workflow — it makes recommendations based on available benchmarks, and leverages the Capacity Analyzer and Performance Analyzer to fill gaps and expand the search surface beyond what benchmarks alone can cover. The Capacity Analyzer and Performance Analyzer are also useful on their own for deeper analysis in their respective areas, each with its own UI page, CLI command, and API endpoint.

### Key Features

> - **Conversational Requirements Gathering** - Describe your use case in natural language

Reviewer 3: ```suggestionAI
- **Conversational Requirements Gathering** - Describe your AI-powered use case in natural language
```

- **GPU Memory Estimation** - Calculate model weight memory, KV cache, activation overhead, and minimum GPU requirements for any HuggingFace model
- **Performance Estimation** - Estimate TTFT, ITL, and throughput across GPU types without running benchmarks
- **SLO-Driven Recommendations** - Get optimal model + GPU configurations backed by real benchmarks, with estimated performance fallback
- **Multi-Criteria Ranking** - Score configurations on accuracy, price, latency, and complexity with 5 ranked views
- **What-If Analysis** - Explore alternatives and compare cost vs. latency tradeoffs
- **One-Click Deployment** - Generate production-ready KServe/vLLM YAML and deploy to Kubernetes
- **Performance Monitoring** - Track actual deployment status and test inference in real-time
- **CLI and API** - Use the `planner` CLI or REST API for programmatic access to all capabilities
- **GPU-Free Development** - vLLM simulator enables local testing without GPU hardware

### How It Works

**Planner flow:**

1. **Extract Intent** - LLM-powered analysis converts your description into structured requirements

> 2. **Map to Traffic Profile** - Match use case to one of 4 GuideLLM benchmark configurations

Reviewer: ```suggestion
2. **Map to Traffic Profile** - Match use case to one of 4 [GuideLLM](https://github.com/vllm-project/guidellm) benchmark configurations
```

> 3. **Set SLO Targets** - Auto-generate TTFT (p95), ITL (p95), and E2E (p95) targets based on experience class

Reviewer: ```suggestion
3. **Set SLO Targets** - Auto-generate TTFT, ITL, and E2E targets based on experience class
```
P95 might be the default now, but we may be more flexible in the future.

> 4. **Query Benchmarks** - Exact match on (model, GPU, traffic profile) from PostgreSQL database

Reviewer: ```suggestion
4. **Query Benchmarks** - Exact match on (model, GPU, traffic profile) from benchmark database
```
PostgreSQL for now, but more generic phrasing allows this to change and keep this statement accurate.

5. **Estimate if Missing** - For models/GPUs without benchmarks, generate estimated performance via Performance Analyzer

> 6. **Filter by SLOs** - Find configurations meeting all p95 latency targets

Reviewer: ```suggestion
6. **Filter by SLOs** - Find configurations meeting all latency and throughput targets
```
Throughput can also be an SLO

Reviewer 2: 2 things.  

1. It could.  I went back and forth about whether to include "throughput" as an SLO or traffic profile in the UI.  I ended up including it with the traffic profile.
2. We don't filter on throughput.  The latency numbers are in part a function of the offered request rate.  So, we find the request rates that meet the SLOs (if any) and then compute the replicas needed to meet the throughput requirements.

I'll replace "Filter by SLOs" with "Filter by latency" to make it more clear.

7. **Plan Capacity** - Calculate required replicas based on throughput requirements
8. **Generate & Deploy** - Create validated Kubernetes YAML and deploy to local or production clusters
9. **Monitor & Validate** - Track deployment status and test inference endpoints

**Capacity analysis flow:**

1. **Fetch Model Config** - Retrieve architecture details from HuggingFace (layers, heads, quantization)

> 2. **Calculate Memory** - Model weights, KV cache per token, activation memory, CUDA/system overhead

Reviewer: ```suggestion
2. **Calculate Memory** - Model weights, KV cache per token, activation memory, system overhead
```

3. **Evaluate Hardware Fit** - Determine valid tensor parallelism values and whether the model fits on a given GPU
4. **Report Capacity** - Maximum context length, concurrent requests, KV cache blocks for the configuration

**Performance analysis flow:**

1. **Select GPUs** - Choose from a catalog of GPU types (H100, A100, L40, etc.) or evaluate all

> 2. **Estimate Performance** - Estimate TTFT, ITL, and throughput per GPU (currently via BentoML roofline model, with additional backends planned)

Reviewer: ```suggestion
2. **Estimate Performance** - Estimate TTFT, ITL, and throughput per GPU (currently via [BentoML LLM Optimizer](https://github.com/bentoml/llm-optimizer) roofline model, with additional backends planned)
```

3. **Apply Constraints** - Filter by max TTFT, max ITL, max latency, and GPU count limits
4. **Rank Results** - Compare qualifying GPUs sorted by performance with cost details

## Prerequisites

**Required before running `make setup`:**

- **macOS or Linux** (Windows via WSL2)

> - **Docker Desktop** (must be running)

Reviewer 3: Is docker desktop required? I thought just Docker or Podman would work

Author: Docker or Podman should do. I'll fix that.

- **[Logging Guide](docs/LOGGING.md)** - Logging system and debugging
- **[Claude Code Guidance](CLAUDE.md)** - AI assistant instructions for contributors

## Future Enhancements

> 1. **Prefill/Decode Disaggregation** - Support P/D disaggregation as a first-class deployment topology

Reviewer 3: I would also add exposing llm-d stack level configuration like routing in addition to PD and finer-grained vLLM params

2. **vLLM Parameter Search** - Expand configuration search to include serving-stack knobs (batch size, scheduling policy, memory utilization)
3. **Feedback Loops** - Connect deployment metrics back to estimation models to improve accuracy over time
4. **Dynamic Workload Adaptation** - Detect traffic shifts and trigger configuration re-evaluation
5. **Production GPU Validation** - End-to-end testing with real GPU clusters
6. **Statistical Traffic Models** - Full distributions (not just point estimates)

> 7. **Multi-Dimensional Benchmarks** - Concurrency, batching, KV cache effects

Reviewer: A lot of these future items cross into other SIGs, do we really expect to support all this in planner? Maybe some level of support will be added, but these statements sound pretty strong as if planner will be the tool to do all of this for you.

Reviewer 2: I would take 3 and 4 off the table for now.  For #5, I've done very limited testing with real GPUs in real kubernetes cluster and want to do more.  But, we don't need that in the README.  I'm fine with deleting 6 &7 too.  We don't need to call them out as a top priority just yet.
